### PR TITLE
feat: `SLOAD` optimized route and price

### PR DESCRIPTION
**Pull request by xenide**
## Currently 
best case scenario:
- simple route
- num of SLOADs = 4
  - 1 for the existence of route
  - 1 for the length of the route array + first item  
  - 1 for the second item of the array 
  - 1 for the price cache

worst case scenario:
- 3 hop route 
- num of SLOADs = 8 
  - 1 for route query
  - 1 for length of route array + first item 
  - 1 for second item of array
  - 1 for third item of array
  - 1 for fourth item of array 
  - 3 for the price cache for 3 hops  

## Implemented solution
best case scenario:
- simple route
- num of SLOADs = 1 
  - `keccak256` token0-token1 and then loads 1 storage slot
  - reads price from that storage slot

Worst case scenario: 
- 3 hop route 
- num of SLOADs = 6
  - 3 for reading route
  - 3 for reading simple prices 

## Potential Further improvements 

1. we can split some addresses over two words 
   -  i.e. storing 11 bytes in the previous word and then 9 bytes in the next word
   - e.g. 3 hop route (total 63 bytes of data) 
     - instead of reading 3 slots for the route, we only have to read 2 slots 
     - so we save 1 SLOAD
2. Storing the decimal multipliers in the unused bits in the slot
   - if we do this then we won't be able to do improvement no.1

## Alternatives tried 

1. Tried using the following: 

```solidity
mapping(address token0 => mapping(address token1 => bytes wtv))
```

Doesn't work because the the `bytes` will track length. So that's one more SLOAD

2. also tried 

```solidity
mapping(address token0 => mapping(address token1 => bytes wtv))[0:32]
```

doesn't even compile

### CODE REVIEW

File: src/ReservoirPriceOracle.sol [hunk 1]
```diff
@@ -353,40 +456,87 @@ contract ReservoirPriceOracle is IPriceOracle, IReservoirPriceOracle, Owned(msg.
...
-                    lSubroute[0] = lLowerToken;
-                    lSubroute[1] = lHigherToken;
+        // simple route
+        // gas: can store aRoute.length as a local variable
+        if (aRoute.length == 2) {
+            assembly {
+                let data := shl(248, 0x01) // 0x01 in the uppermost byte
+                sstore(lSlot, data)
+            }
+            // update price for simple routes
```
Comment (OliverNChalk) [inline src/ReservoirPriceOracle.sol]: Is this unfinished?

File: src/ReservoirPriceOracle.sol [hunk 2]
```diff
@@ -353,40 +456,87 @@ contract ReservoirPriceOracle is IPriceOracle, IReservoirPriceOracle, Owned(msg.
...
+                assembly {
+                    lRead := sload(lIntermediateRouteSlot)
+                }
+                if (lRead == bytes32(0)) {
+                    address[] memory lIntermediateRoute = new address[](2);
+                    lIntermediateRoute[0] = lLowerToken;
+                    lIntermediateRoute[1] = lHigherToken;
+                    setRoute(lLowerToken, lHigherToken, lIntermediateRoute);
                 }
             }
```
Comment (OliverNChalk) [inline src/ReservoirPriceOracle.sol]: Please comment what is happening here, its very hard to follow

File: src/ReservoirPriceOracle.sol [hunk 3]
```diff
@@ -353,40 +456,87 @@ contract ReservoirPriceOracle is IPriceOracle, IReservoirPriceOracle, Owned(msg.
...
+            bytes32 lRead;
+            assembly {
+                lRead := sload(lIntermediateRouteSlot)
+            }
+            if (lRead == bytes32(0)) {
+                address[] memory lIntermediateRoute = new address[](2);
+                lIntermediateRoute[0] = lLowerToken;
+                lIntermediateRoute[1] = lHigherToken;
+                setRoute(lLowerToken, lHigherToken, lIntermediateRoute);
+            }
```
Comment (OliverNChalk) [inline src/ReservoirPriceOracle.sol]: Full sentences explaining whats happening

**Review by OliverNChalk — CHANGES_REQUESTED** [review]
Look directionally correcy, would be good:

- to get some proper commenting of the more complex code (full sentence explanations of what it should be doing and why)

File: src/ReservoirPriceOracle.sol (hunk 1 above)
Comment (xenide) [reply src/ReservoirPriceOracle.sol]: I'm a lil conflicted for this one. Do you think we should call `updatePrice` when setting the route? 

On one hand, it's good that we update the price right away, so that anyone who queries the oracle right after the route setting will obtain a valid price. This can be easily mitigated by separating calling `updatePrice`. 

On the other hand, it makes the function less 'clean'. It's doing more than it's supposed to. 

I'm leaning towards NOT updating the price in when setting the route.

File: src/ReservoirPriceOracle.sol (hunk 2 above)
Comment (xenide) [reply src/ReservoirPriceOracle.sol]: refactored and added comments

File: src/ReservoirPriceOracle.sol (hunk 1 above)
Comment (OliverNChalk) [reply src/ReservoirPriceOracle.sol]: Agree not updating when setting

File: src/ReservoirPriceOracle.sol [hunk 4]
```diff
@@ -141,11 +146,12 @@ contract ReservoirPriceOracle is IPriceOracle, IReservoirPriceOracle, Owned(msg.
     /// for priceCache[aTokenA][aTokenB] but instead the prices of its constituent simple routes will be written.
     /// @param aTokenA Address of one of the tokens for the price update. Does not have to be less than address of aTokenB
     /// @param aTokenB Address of one of the tokens for the price update. Does not have to be greater than address of aTokenA
-    /// @param aRewardRecipient The beneficiary of the reward. Must implement the receive function if is a smart contract address
-    function updatePrice(address aTokenA, address aTokenB, address aRewardRecipient) external nonReentrant {
+    /// @param aRewardRecipient The beneficiary of the reward. Must implement the receive function if is a smart
+    /// contract address. Set to address(0) if not seeking a reward
```
Comment (OliverNChalk) [inline src/ReservoirPriceOracle.sol]: Would change "must implement the receive function" to "must be able to receive ether" because im not sure if vyper calls it receive & I first thought this meant there was an interface to implement for receiving rewards.

File: src/ReservoirPriceOracle.sol [hunk 5]
```diff
@@ -350,43 +483,89 @@ contract ReservoirPriceOracle is IPriceOracle, IReservoirPriceOracle, Owned(msg.
...
+        else {
+            address lSecondToken = aRoute[1];
+            address lThirdToken = aRoute[2];
+            // Set the uppermost byte of lFirstWord to FLAG_COMPOSITE_NEXT for intermediate hops
+            bytes32 lFirstWord = FlagsLib.FLAG_COMPOSITE_NEXT << 248;
+            // Move the address to start on the 2nd byte.
+            bytes32 lSecondTokenData = bytes32(bytes20(lSecondToken)) >> 8;
+            bytes32 lThirdTokenFirst10Bytes = bytes32(bytes20(lThirdToken)) >> 176;
+
+            // Trim away the first 10 bytes since we only want the last 10 bytes.
```
Comment (OliverNChalk) [inline src/ReservoirPriceOracle.sol:521]: is this comment correct?

File: src/ReservoirPriceOracle.sol [hunk 6]
```diff
@@ -350,43 +483,89 @@ contract ReservoirPriceOracle is IPriceOracle, IReservoirPriceOracle, Owned(msg.
...
+                lFirstWord =
+                    lFirstWord | lSecondTokenData | FlagsLib.FLAG_COMPOSITE_NEXT << 80 | lThirdTokenFirst10Bytes;
+
+                bytes32 lFourthTokenData = bytes32(bytes20(aToken1)) >> 88;
+                lSecondWord = lThirdTokenSecond10Bytes | FlagsLib.FLAG_COMPOSITE_END << 168 | lFourthTokenData;
+
+                _checkAndPopulateIntermediateRoute(lThirdToken, aToken1);
+            }
+            _checkAndPopulateIntermediateRoute(aToken0, lSecondToken);
+            _checkAndPopulateIntermediateRoute(lSecondToken, lThirdToken);
```
Comment (OliverNChalk) [inline src/ReservoirPriceOracle.sol:541]: I'm confused by this, assume the route:

A -> B -> C

Are AB and BC the intermediate routes? Why do we need to store them too?

File: src/ReservoirPriceOracle.sol (hunk 4 above)
Comment (xenide) [reply src/ReservoirPriceOracle.sol]: you're right, vyper doesn't seem to have the `receive` function. Changed comments

File: src/ReservoirPriceOracle.sol (hunk 5 above)
Comment (xenide) [reply src/ReservoirPriceOracle.sol:521]: yes. When we convert address => bytes20 we don't lose any address data. Then when we convert bytes20 => bytes32 it pads right. So when address converted into bytes32 it occupies the top 20 bytes. Shifting left by 80 bits will get rid of the top 10 bytes of the address, leaving the bottom 10 bytes of the address on the top 10 bytes of the bytes32

File: src/ReservoirPriceOracle.sol (hunk 6 above)
Comment (xenide) [reply src/ReservoirPriceOracle.sol:541]: as explained in discord, resolving

**Review by OliverNChalk — APPROVED** [review]
I think the impl/design is good, we probably want some fuzz tests but that can be added after merging if you prefer:

- Generate a random 2 hop route. Generate 2 random prices & insert. Confirm the quote matches a manual compute of expected quote.
- Generate a random 3 hop route. Generate 3 random prices & insert. Confirm the quote matches a manual compute of expected quote.
- Generate a random 4 hop route. Generate 4 random prices & insert. Confirm the quote matches a manual compute of expected quote.
